src/main/java/htsjdk/samtools/SamReaderFactory.java

```diff
@@ -311,7 +311,24 @@ public SamReader open(final SamInputResource resource) {
                     if (SamStreams.isBAMFile(bufferedStream)) {
                         if (sourceFile == null || !sourceFile.isFile()) {
                             // Handle case in which file is a named pipe, e.g. /dev/stdin or created by mkfifo
-                            primitiveSamReader = new BAMFileReader(bufferedStream, indexFile, false, asynchronousIO, validationStringency, this.samRecordFactory);
+                            if (indexFile!=null) {
+                                primitiveSamReader = new BAMFileReader(bufferedStream, indexFile, false, asynchronousIO, validationStringency, this.samRecordFactory);
+                            } else {
+                                final SeekableStream indexSeekable = indexMaybe == null ? null : indexMaybe.asUnbufferedSeekableStream();
+                                // do not close bufferedStream, it's the same stream we're getting here.
+                                SeekableStream sourceSeekable = data.asUnbufferedSeekableStream();
+                                if (null != sourceSeekable && null != indexSeekable) {
+                                    // let's make sure the user can seek in this file.
+                                    // need to return to the beginning because it's the same stream we used earlier
+                                    // and read a bit from.
+                                    sourceSeekable.seek(0);
```

Comment:
Can you explain why we don't have to seek to the beginning in the other places where `bufferedStream` are used (assuming `bufferedStream` and `sourceSeekable` are the same underlying stream)?